# Report 
Group 5 - Product in consideration Diacetyl (mmol/gDW/h)\
Guilherme Lobo\
Karyna Lysenko\
Rodrigo Esperança

Biologia de Sistemas - Universidade do Minho\
dezembro de 2022

### Documentação
https://cobrapy.readthedocs.io/en/latest/

In [2]:
import cobra
from cobra.io import read_sbml_model
from cobra.flux_analysis import flux_variability_analysis as fva
import pandas
import mewpy
from mewpy.simulation import get_simulator
from mewpy.omics import ExpressionSet
from mewpy.omics import GIMME

In [3]:
model = read_sbml_model("sth_novo.xml")
#model

#### To find out the Diacetyl metabolite id:

In [3]:
for i in model.metabolites:
    if i.name == "Diacetyl_":
        print(i.id)

diact_c
diact_e


In [4]:
diact_c=model.metabolites.get_by_id("diact_c")
#diact_c
diact_c=model.metabolites.get_by_id("diact_e")
#diact_c
DIACTt= model.reactions.get_by_id("DIACTt")
#DIACTt
casp= model.reactions.get_by_id("EX_caspep_e")
#casp
EX_diact_e= model.reactions.get_by_id("EX_diact_e")
#EX_diact_e

## Chemical defined medium (CDM)
### Function objetive: maximization of biomass production.

##### 1 a) What is the wild-type production of the compound?


In order to have a general idea of the flux of biomass_STR and of the exchange reactions we applied the summary() function and noted the biomass production is 1.042 1/h.

In [5]:
model.summary()

In [4]:
model.optimize()["EX_diact_e"]

0.0

Conclusion: The wild-type production of the Diacetyl is 0.0 mmol/gDW/h.

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [8]:
model_fva=fva(model,fraction_of_optimum=1)
model_fva.loc["EX_diact_e"]

minimum    0.000000e+00
maximum    5.098144e-13
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very close to each other (maximum is almost 0.0), we can conclude that the solution is robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [5]:
model_1c = read_sbml_model("sth_novo.xml")

In [10]:
model_1c.summary(fva=0.20)

In [11]:
model_1c.optimize()["EX_diact_e"]

-0.0

Conclusion: The Diacetyl production is still 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [12]:
from cobra.flux_analysis import (single_gene_deletion, double_gene_deletion)

In [6]:
#ranking in order of Diacetyl production
list_variants_D=[]
for i in model.genes:
    with model:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        save_lactose_optimize=float(tmp["EX_lcts_e"])
        gene=str(i)
        objective=model.slim_optimize()
        variant = (save_diact_optimize, gene, objective,save_lactose_optimize)
        list_variants_D.append(variant)

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

In [11]:
sorted(list_variants_D, reverse=True)

[(1.4793217485434982e-29, 'CH8_2083', 3.9470128746759845e-13, -25.11),
 (0.0, 'CH8_2151', 0.0, -25.11),
 (0.0, 'CH8_2150', 1.0391380554050986, -25.11),
 (0.0, 'CH8_2144', 0.0, -25.11),
 (0.0, 'CH8_2143', 0.0, -25.11),
 (0.0, 'CH8_2142', 0.0, -25.11),
 (0.0, 'CH8_2141', 0.0, -25.11),
 (0.0, 'CH8_2108', 0.0, -25.11),
 (0.0, 'CH8_2106', 0.0, -25.11),
 (0.0, 'CH8_2100', 1.0422533807364607, -25.11),
 (0.0, 'CH8_2056', 7.77237761949506e-15, -25.11),
 (0.0, 'CH8_2042', nan, -25.11),
 (0.0, 'CH8_2019', 1.0422533807364571, -25.11),
 (0.0, 'CH8_2001', 1.0422533807364098, -25.11),
 (0.0, 'CH8_1991', 1.0422533807364098, -25.11),
 (0.0, 'CH8_1988', 1.0422533807364607, -25.11),
 (0.0, 'CH8_1987', 1.0422533807364205, -25.11),
 (0.0, 'CH8_1986', 1.0422533807364205, -25.11),
 (0.0, 'CH8_1985', 1.0422533807364205, -25.11),
 (0.0, 'CH8_1984', 1.0422533807364098, -25.11),
 (0.0, 'CH8_1983', 0.0, -25.11),
 (0.0, 'CH8_1982', 1.04225338073645, -25.11),
 (0.0, 'CH8_1964', 1.0186155578520841, -25.11),
 (0.0, '

Como a taxa de crescimento para os 4 melhores( em termos de produção do composto diacetil) tem valor 0 ao fazer o cálculo do BCPI, este vai ter valor de 0 1/H. 

As the growth performance for the 4 best values of Diacetyl production is 0, the calculaton of BCPY is also 0 1/h.

In [72]:
#alternativa para o sorted em função da produção do composto diacetil
def Sort(sub_li):
    sub_li.sort(key = lambda x: x[0])
    return sub_li

sub_li =list_variants_D
Sort(sub_li)

[(-0.0, 'CH8_0371', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_1345', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0370', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_1344', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0372', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0555', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0124', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0263', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_1982', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0662', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0280', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_1211', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0680', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0149', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0947', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_1796', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0069', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0070', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0354', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0314', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0313', 

In [73]:
#alternativa para o sorted em função dos genes
def Sort(sub_li):
    sub_li.sort(key = lambda x: x[1])
    return sub_li

sub_li =list_variants_D
Sort(sub_li)

[(-0.0, 'CH8_0012', 1.022534141822745, -25.11),
 (-0.0, 'CH8_0051', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0054', 1.042253380736514, -25.11),
 (-0.0, 'CH8_0058', 1.0422533807364442, -25.11),
 (-0.0, 'CH8_0059', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0060', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0061', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0062', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0063', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0068', 1.0422533807363843, -25.11),
 (-0.0, 'CH8_0069', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0070', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0074', 0.0, -25.11),
 (-0.0, 'CH8_0124', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0127', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0128', 0.0, -25.11),
 (-0.0, 'CH8_0137', 1.0422533807367802, -25.11),
 (-0.0, 'CH8_0149', 1.0422533807363938, -25.11),
 (-0.0, 'CH8_0154', 1.042253380736514, -25.11),
 (-0.0, 'CH8_0155', 1.0422533807367802, -25.11),
 (-0.0, 'CH8_0156', 1.0422533807364305, -25.11),
 (-0

Relativamente ao cálculo do BCPI quando temos em consideração os maiores 4 valores de taxa de crescimento podemos verificar que vai ser 0, uma vez que, os valores diferentes de zero obtidos para produção de diacetil não correspondem aos valores máximos da taxa de crescimento. Portanto, a produção do nosso produto será sempre, neste caso em particular, 0 ou na.

In [9]:
import numpy as np
cleanedList = []
for y in list_variants_D:
    if not np.isnan(y[2]):
        cleanedList.append(y)

In [10]:
#alternativa para o sorted em função da taxa de crescimento
#quando queremos fazer o rank da taxa de crescimento o sort não funciona
def Sort(sub_li):
    sub_li.sort(key = lambda x: x[2])
    return sub_li

sub_li =cleanedList
Sort(sub_li)

[(0.0, 'CH8_0352', -1.1271542576555125e-13, -25.11),
 (0.0, 'CH8_1177', -3.91275443631472e-14, -25.11),
 (0.0, 'CH8_1893', -1.0817328056077313e-14, -25.11),
 (0.0, 'CH8_0423', -1.8153593930858235e-16, -25.11),
 (0.0, 'CH8_0564', 0.0, -25.11),
 (0.0, 'CH8_0565', 0.0, -25.11),
 (0.0, 'CH8_0854', 0.0, -25.11),
 (0.0, 'CH8_0856', 0.0, -25.11),
 (0.0, 'CH8_0855', 0.0, -25.11),
 (0.0, 'CH8_0857', 0.0, -25.11),
 (0.0, 'CH8_1227', 0.0, -25.11),
 (0.0, 'CH8_1370', 0.0, -25.11),
 (0.0, 'CH8_1372', 0.0, -25.11),
 (0.0, 'CH8_1900', 0.0, -25.11),
 (0.0, 'CH8_1717', 0.0, -25.11),
 (0.0, 'CH8_1718', 0.0, -25.11),
 (0.0, 'CH8_1716', 0.0, -25.11),
 (0.0, 'CH8_0074', 0.0, -25.11),
 (0.0, 'CH8_0128', 0.0, -25.11),
 (0.0, 'CH8_0178', 0.0, -25.11),
 (0.0, 'CH8_0247', 0.0, -25.11),
 (0.0, 'CH8_0248', 0.0, -25.11),
 (0.0, 'CH8_0250', 0.0, -25.11),
 (0.0, 'CH8_0271', 0.0, -25.11),
 (0.0, 'CH8_0277', 0.0, -25.11),
 (0.0, 'CH8_0289', 0.0, -25.11),
 (0.0, 'CH8_0291', 0.0, -25.11),
 (0.0, 'CH8_0304', 0.0, -25.11)

In [ ]:
#

Conclusion: 

## Milk medium (MM)
### Function objetive: minimization of casein consumption while constraining the growth rate to 0.64 h-1

In [12]:
model_milk = read_sbml_model("sth_novo.xml")

In [13]:
model_milk.objective = 'EX_caspep_e'

model_milk.reactions.biomass_STR.lower_bound=0.64
model_milk.reactions.biomass_STR.upper_bound=0.64

#change the conditions of the solution
model_milk.reactions.EX_lac_L_e.lower_bound=30
model_milk.reactions.EX_lac_L_e.upper_bound=40.8

model_milk.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk.reactions.EX_gal_e.lower_bound=0
model_milk.reactions.EX_gal_e.upper_bound=1000

model_milk.reactions.EX_glc_D_e.lower_bound=2
model_milk.reactions.EX_glc_D_e.upper_bound=2.5

model_milk.reactions.EX_caspep_e.lower_bound=-1000
model_milk.reactions.EX_caspep_e.upper_bound=1000

model_milk.reactions.EX_mpept_e.lower_bound=0
model_milk.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk.reactions.EX_ala_D_e.lower_bound = 0
model_milk.reactions.EX_ala_D_e.upper_bound = 1000
model_milk.reactions.EX_ala_L_e.lower_bound = 0
model_milk.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk.reactions.EX_arg_L_e.lower_bound = 0
model_milk.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk.reactions.EX_asn_L_e.lower_bound = 0
model_milk.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk.reactions.EX_asp_L_e.lower_bound = 0
model_milk.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk.reactions.EX_cys_L_e.lower_bound = 0
model_milk.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk.reactions.EX_gln_L_e.lower_bound = 0
model_milk.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk.reactions.EX_glu_L_e.lower_bound=0
model_milk.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk.reactions.EX_gly_e.lower_bound=0
model_milk.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk.reactions.EX_his_L_e.lower_bound=0
model_milk.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk.reactions.EX_ile_L_e.lower_bound=0
model_milk.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk.reactions.EX_leu_L_e.lower_bound=0
model_milk.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk.reactions.EX_lys_L_e.lower_bound=0
model_milk.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk.reactions.EX_met_L_e.lower_bound=0
model_milk.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk.reactions.EX_phe_L_e.lower_bound=0
model_milk.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk.reactions.EX_pro_L_e.lower_bound=0
model_milk.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk.reactions.EX_ser_D_e.lower_bound=0
model_milk.reactions.EX_ser_D_e.upper_bound=1000
model_milk.reactions.EX_ser_L_e.lower_bound=0
model_milk.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk.reactions.EX_thr_L_e.lower_bound=0
model_milk.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk.reactions.EX_trp_L_e.lower_bound=0
model_milk.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk.reactions.EX_tyr_L_e.lower_bound=0
model_milk.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk.reactions.EX_val_L_e.lower_bound=0
model_milk.reactions.EX_val_L_e.upper_bound=1000

#### 1 a) What is the wild-type production of the compound?

In [15]:
model_milk.summary()

In [16]:
model_milk.optimize()["EX_diact_e"]

1.346139990204063

Conclusion: The wild-type production of the Diacetyl is 1.346 mmol/gDW/h. 

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [36]:
model_milk_fva=fva(model_milk,fraction_of_optimum=1)
model_milk_fva.loc["EX_diact_e"] 

minimum    0.00000
maximum    7.05441
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very distant to each other, we can conclude that the solution is not robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [37]:
model_milk_20 = read_sbml_model("sth_novo.xml")

In [38]:
model_milk_20.objective = 'EX_caspep_e'

model_milk_20.reactions.biomass_STR.lower_bound=0.64
model_milk_20.reactions.biomass_STR.upper_bound=0.64

#change the conditions of the solution
model_milk_20.reactions.EX_lac_L_e.lower_bound=30
model_milk_20.reactions.EX_lac_L_e.upper_bound=40.8

model_milk_20.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk_20.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk_20.reactions.EX_gal_e.lower_bound=0
model_milk_20.reactions.EX_gal_e.upper_bound=1000

model_milk_20.reactions.EX_glc_D_e.lower_bound=2
model_milk_20.reactions.EX_glc_D_e.upper_bound=2.5

model_milk_20.reactions.EX_caspep_e.lower_bound=-1000
model_milk_20.reactions.EX_caspep_e.upper_bound=1000

model_milk_20.reactions.EX_mpept_e.lower_bound=0
model_milk_20.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk_20.reactions.EX_ala_D_e.lower_bound = 0
model_milk_20.reactions.EX_ala_D_e.upper_bound = 1000
model_milk_20.reactions.EX_ala_L_e.lower_bound = 0
model_milk_20.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk_20.reactions.EX_arg_L_e.lower_bound = 0
model_milk_20.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk_20.reactions.EX_asn_L_e.lower_bound = 0
model_milk_20.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk_20.reactions.EX_asp_L_e.lower_bound = 0
model_milk_20.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk_20.reactions.EX_cys_L_e.lower_bound = 0
model_milk_20.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk_20.reactions.EX_gln_L_e.lower_bound = 0
model_milk_20.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk_20.reactions.EX_glu_L_e.lower_bound=0
model_milk_20.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk_20.reactions.EX_gly_e.lower_bound=0
model_milk_20.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk_20.reactions.EX_his_L_e.lower_bound=0
model_milk_20.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk_20.reactions.EX_ile_L_e.lower_bound=0
model_milk_20.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk_20.reactions.EX_leu_L_e.lower_bound=0
model_milk_20.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk_20.reactions.EX_lys_L_e.lower_bound=0
model_milk_20.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk_20.reactions.EX_met_L_e.lower_bound=0
model_milk_20.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk_20.reactions.EX_phe_L_e.lower_bound=0
model_milk_20.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk_20.reactions.EX_pro_L_e.lower_bound=0
model_milk_20.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk_20.reactions.EX_ser_D_e.lower_bound=0
model_milk_20.reactions.EX_ser_D_e.upper_bound=1000
model_milk_20.reactions.EX_ser_L_e.lower_bound=0
model_milk_20.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk_20.reactions.EX_thr_L_e.lower_bound=0
model_milk_20.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk_20.reactions.EX_trp_L_e.lower_bound=0
model_milk_20.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk_20.reactions.EX_tyr_L_e.lower_bound=0
model_milk_20.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk_20.reactions.EX_val_L_e.lower_bound=0
model_milk_20.reactions.EX_val_L_e.upper_bound=1000

In [39]:
model_milk_20.reactions.biomass_STR.lower_bound=0.64*0.2
model_milk_20.reactions.biomass_STR.upper_bound=0.64*0.2 

In [17]:
model_milk_20.optimize()["EX_diact_e"] 

NameError: name 'model_milk_20' is not defined

Conclusion: The Diacetyl production is 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [19]:
#ranking in order of Diacetyl production and growth performance
list_variants_D_MM=[]
for i in model_milk.genes:
    with model_milk:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model_milk.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        save_lactose_optimize=float(tmp["EX_lcts_e"])
        gene=str(i)
        objective=model_milk.slim_optimize()
        variant = (save_diact_optimize, gene, objective,save_lactose_optimize)
        list_variants_D_MM.append(variant)

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

In [20]:
#ranking in order of Diacetyl production
sorted(list_variants_D_MM, reverse=True)

[(6.9877614079999395, 'CH8_0368', nan, -24.773848661333336),
 (6.9877614079999395, 'CH8_0352', nan, -24.773848661333336),
 (6.935134208000038, 'CH8_1177', nan, -24.721221461333347),
 (6.935134208000035, 'CH8_1893', nan, -24.721221461333336),
 (6.935134207999991, 'CH8_1718', nan, -24.721221461333347),
 (6.935134207999991, 'CH8_1717', nan, -24.721221461333347),
 (6.935134207999991, 'CH8_1716', nan, -24.721221461333347),
 (6.935134207999991, 'CH8_1680', -0.008768319999999996, -24.721221461333347),
 (6.93513420799999, 'CH8_1681', -0.008768319999999996, -24.72122146133336),
 (6.770437551999983, 'CH8_0428', nan, -24.55698912533335),
 (6.770437551999983, 'CH8_0423', nan, -24.55698912533335),
 (6.48722459036736, 'CH8_0332', -0.007148042448979599, -24.319578971863958),
 (6.487224590367349, 'CH8_0336', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0335', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0334', -0.007148042448979599, -24.3195789718639

In [21]:
import numpy as np
cleanedListMM = []
for y in list_variants_D_MM:
    if not np.isnan(y[2]):
        cleanedListMM.append(y)

In [22]:
#ranking in order of growth performance
def Sort(sub_li):
    sub_li.sort(key = lambda x: x[2])
    return sub_li
Sort(cleanedListMM)

[(0.0, 'CH8_1881', -0.30386549668161317, -25.11),
 (0.0, 'CH8_1885', -0.30386549668161317, -25.11),
 (0.0, 'CH8_1343', -0.01927689846153839, -23.649730436923047),
 (6.93513420799999, 'CH8_1681', -0.008768319999999996, -24.72122146133336),
 (6.935134207999991, 'CH8_1680', -0.008768319999999996, -24.721221461333347),
 (4.45307392, 'CH8_1677', -0.008768319999999996, -22.6),
 (4.453073920000005, 'CH8_1678', -0.008768319999999996, -22.6),
 (4.453073920000005, 'CH8_1679', -0.008768319999999996, -22.6),
 (0.0, 'CH8_1728', -0.008199956078431374, -22.6),
 (6.48722459036736, 'CH8_0332', -0.007148042448979599, -24.319578971863958),
 (6.487224590367349, 'CH8_0333', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0334', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0335', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0336', -0.007148042448979599, -24.319578971863955),
 (6.487224590367349, 'CH8_0330', -0.007148042448979599, -24

Conclusion: The last output is a list of tuples where 1st value is the Diacetyl production, the 2nd is the gene that was deleted and the 3rd is the grouth performance. The list was sorted by the 1st value. We can note several things:
-the deletion of some genes turns the system infeasable -> "nan";

-the deletion of some genes, such as CH8_1680, CH8_1681, CH8_0428, etc., increases the production of Diacetyl but in terms of maximization of the objective function it gets worse, as the 3rd value gets higher in module, so the consumption of EX_caspep_e is higher in those cases;

-there are some other cases of gene deleltion, such as CH8_0316, CH8_1892, etc., qhere the flux of the objective function is the same before any deletions (-0.0030233600000000157 mmol/gDW/h) but the production of the Diacetyl is higher, around 3.18 mmol/gDW/h.

-there are some gene deletions, where there is no production of Diacetyl.


#### 2) The authors of the model obtained gene expression data (RNA-seq) for this strain in CDM and milk medium (“gene_expression.xlsx”) in anaerobic conditions. For this exercise, consider the maximization of the growth rate as the objective for both media. For milk medium, set the uptake of casein as 0.004 mmol/gDW/h.

#### a) Integrate the expression data for both conditions using the GIMME algorithm. Compare the production of your compound in both conditions, as well as the flux of its production pathway(s).

In [45]:
gene_expression = pandas.read_excel("gene_expression.xlsx")
gene_expression

ID         CDM        MILK
0     CH8_0001  198.412656  217.468245
1     CH8_0002  323.207616  319.916967
2     CH8_0003   53.441435   58.719433
3     CH8_0004  333.943347  349.360850
4     CH8_0005   40.002946   56.077228
...        ...         ...         ...
2156  CH8_2157    0.000000    1.674210
2157  CH8_2158  142.603093  163.920837
2158  CH8_2159  159.853040  169.367615
2159  CH8_2160  643.218883  687.317443
2160  CH8_2161   52.616722   80.981196

[2161 rows x 3 columns]

## GIMME- Chemical defined medium (CDM)

In [76]:
envcond = {
"EX_o2_e":(0,0) 
 }
#Usa-se este comando para insturir o meio anaerobio e não se coloca mais nada porque o meio ja esta defenido no modelo

simul = get_simulator(model, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["CDM"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g = GIMME(simul, expr, parsimonious=True)

print(g)
print()
print(model.optimize()["EX_diact_e"])
print(model.optimize()["EX_lcts_e"])

objective: biomass_STR
Status: OPTIMAL
Constraints: OrderedDict([('EX_o2_e', (0, 0)), ('biomass_STR', (0.9380280426628942, inf))])
Method:None

-0.0
-25.11


BCPY=0 

## GIMME- Milk medium (MM)

In [47]:
model_GIMME_MM = read_sbml_model("sth_novo.xml")


#change the conditions of the solution
model_GIMME_MM.reactions.EX_lac_L_e.lower_bound=30
model_GIMME_MM.reactions.EX_lac_L_e.upper_bound=40.8

model_GIMME_MM.reactions.EX_lcts_e.lower_bound= (-25.11)
model_GIMME_MM.reactions.EX_lcts_e.upper_bound= (-22.6)

model_GIMME_MM.reactions.EX_gal_e.lower_bound=0
model_GIMME_MM.reactions.EX_gal_e.upper_bound=1000

model_GIMME_MM.reactions.EX_glc_D_e.lower_bound=2
model_GIMME_MM.reactions.EX_glc_D_e.upper_bound=2.5

model_GIMME_MM.reactions.EX_caspep_e.lower_bound=-0.004
#model_GIMME_MM.reactions.EX_caspep_e.upper_bound=1000

model_GIMME_MM.reactions.EX_mpept_e.lower_bound=0
model_GIMME_MM.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_GIMME_MM.reactions.EX_ala_D_e.lower_bound = 0
model_GIMME_MM.reactions.EX_ala_D_e.upper_bound = 1000
model_GIMME_MM.reactions.EX_ala_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_GIMME_MM.reactions.EX_arg_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_GIMME_MM.reactions.EX_asn_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_GIMME_MM.reactions.EX_asp_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_GIMME_MM.reactions.EX_cys_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_GIMME_MM.reactions.EX_gln_L_e.lower_bound = 0
model_GIMME_MM.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_GIMME_MM.reactions.EX_glu_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_GIMME_MM.reactions.EX_gly_e.lower_bound=0
model_GIMME_MM.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_GIMME_MM.reactions.EX_his_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_GIMME_MM.reactions.EX_ile_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_GIMME_MM.reactions.EX_leu_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_GIMME_MM.reactions.EX_lys_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_GIMME_MM.reactions.EX_met_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_GIMME_MM.reactions.EX_phe_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_GIMME_MM.reactions.EX_pro_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_GIMME_MM.reactions.EX_ser_D_e.lower_bound=0
model_GIMME_MM.reactions.EX_ser_D_e.upper_bound=1000
model_GIMME_MM.reactions.EX_ser_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_GIMME_MM.reactions.EX_thr_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_GIMME_MM.reactions.EX_trp_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_GIMME_MM.reactions.EX_tyr_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_GIMME_MM.reactions.EX_val_L_e.lower_bound=0
model_GIMME_MM.reactions.EX_val_L_e.upper_bound=1000

In [80]:
envcond = {
"EX_o2_e":(0,0) #oxygen exchange
 }

simul = get_simulator(model_GIMME_MM, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["MILK"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g = GIMME(simul, expr, parsimonious=True)
print(g)
print()
print(model_GIMME_MM.optimize()["EX_diact_e"])
print(model_GIMME_MM.optimize()["EX_lcts_e"])

objective: biomass_STR
Status: OPTIMAL
Constraints: OrderedDict([('EX_o2_e', (0, 0)), ('biomass_STR', (0.7620660457239711, inf))])
Method:None

-23.991906548123083


In [23]:
BCPY =(model_GIMME_MM.optimize()["EX_diact_e"]*0.7621)/model_GIMME_MM.optimize()["EX_lcts_e"]
BCPY

NameError: name 'model_GIMME_MM' is not defined

A produção do Diacetil é maior no meio leite

#### b) Compare the output of the algorithm with the results obtained in exercise 1 a).